malloc/malloc.c

```diff
@@ -12,7 +12,22 @@
 */
 char *string_dup(char *src)
 {
-
+    int length = string_length(src);
+    // making memory space for the duplicate
+    char *duplicate = malloc(length);
+    // making a copy of duplicate to do pointer arithmetic with
+    char *duplicate_copy = duplicate;
+    // looping through src until hitting the null char
+    while (*src != '\0')
+    {
+        // mapping src values to duplicate values
+        *duplicate_copy = *src;
+        src++;
+        duplicate_copy++;
+    }
+    // adding null char to end of duplicate string
+    *duplicate_copy = '\0';
+    return duplicate;
```

Comment:
It is interesting that you have a malloc called duplicate and a local variable called duplicate_copy to alter the values in duplicate.  While this might still work, it is probably unnecessary to have duplicate_copy; you can simply alter the values in duplicate using the just the variable duplicate.